src/detect-mqtt-connect-flags.c

```diff
@@ -0,0 +1,424 @@
+    sigmatch_table[DETECT_AL_MQTT_CONNECT_FLAGS].Free  = DetectMQTTConnectFlagsFree;
+    sigmatch_table[DETECT_AL_MQTT_CONNECT_FLAGS].RegisterTests = MQTTConnectFlagsRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex);
+
+    DetectAppLayerInspectEngineRegister("mqtt.connect.flags",
+            ALPROTO_MQTT, SIG_FLAG_TOSERVER, 1,
+            DetectEngineInspectMQTTConnectFlagsGeneric);
+
+    mqtt_connect_flags_id = DetectBufferTypeGetByName("mqtt.connect.flags");
+}
+
+static int DetectEngineInspectMQTTConnectFlagsGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
+                                          f, flags, alstate, txv, tx_id);
+}
+
+/**
+ * \internal
+ * \brief Function to match variable header flags of an MQTT CONNECT Tx
+ *
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param txv     Pointer to the transaction.
+ * \param s       Pointer to the Signature.
+ * \param ctx     Pointer to the sigmatch that we will cast into DetectMQTTConnectFlagsData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectMQTTConnectFlagsMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state,
+                               void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+{
+    const DetectMQTTConnectFlagsData *de = (const DetectMQTTConnectFlagsData *)ctx;
+
+    if (!de)
+        return 0;
+
+    return rs_mqtt_tx_has_connect_flags(txv, de->username, de->password, de->will,
+                                        de->will_retain, de->clean_session);
+   }
+
+/**
+ * \internal
+ * \brief This function is used to parse options passed via mqtt.connect.flags: keyword
+ *
+ * \param rawstr Pointer to the user provided options
+ *
+ * \retval de pointer to DetectMQTTConnectFlagsData on success
+ * \retval NULL on failure
+ */
+static DetectMQTTConnectFlagsData *DetectMQTTConnectFlagsParse(const char *rawstr)
+{
+    DetectMQTTConnectFlagsData *de = NULL;
+#define MAX_SUBSTRINGS 30
+    int ret = 0;
+    int ov[MAX_SUBSTRINGS];
+
+    ret = DetectParsePcreExec(&parse_regex, rawstr, 0, 0, ov, MAX_SUBSTRINGS);
+    if (ret < 1) {
+        SCLogError(SC_ERR_PCRE_MATCH, "invalid flag definition: %s", rawstr);
+        return NULL;
+    }
+
+    de = SCMalloc(sizeof(DetectMQTTConnectFlagsData));
+    if (unlikely(de == NULL))
+        return NULL;
+    de->username = de->password = de->will = DONT_CARE;
+    de->will_retain = de->clean_session = DONT_CARE;
+
+    char copy[strlen(rawstr)+1];
```

Comment:
I do like doing it in Rust, but not keen to enforce that now since its almost more work to do it at this time..  Until a time that the Rust wraps the whole detect glue at least.

Reply:
I guess this can wait then.